### ISSUE TITLE
Don't stringify logged Mojo::Exceptions

### ISSUE THREAD

**Issue by jrubinator** [mojolicious/mojo#1499]
* Mojolicious version: 8.40
* Perl version: 5.18
* Operating system: Centos6

### Short overview of this request
We're using https://metacpan.org/pod/Mojolicious::Plugin::Log::Any to reroute app logs through `Log::Any` to an external bug tracker (Sentry). In 8.24, `Mojo::Log` began stringifying requests earlier (https://metacpan.org/source/SRI/Mojolicious-8.24/lib/Mojo/Log.pm#L72) in a way that prevents us from getting metadetails about what went wrong (such as the stack trace).

### Steps to reproduce the behavior
Logging a `Mojo::Exception` results in premature stringification of the exception when the app log has a context. This occurs by default on exceptions as a combination of
https://metacpan.org/source/SRI/Mojolicious-8.40/lib/Mojo/Log.pm#L72
and https://metacpan.org/source/SRI/Mojolicious-8.40/lib/Mojolicious/Plugin/DefaultHelpers.pm#L176

To see this in action:
```
$ cat log.t
use strict;
use warnings;

use Log::Any::Adapter;
use Test::Mojo;
use Test::More;

{
    package Log::Any::Adapter::Mojo::Test; # this package wants an error object
    use base qw(Log::Any::Adapter::Base);

    use Test::More;

    sub structured {
        my ($self, $level, $category, @log_args) = @_;

        # This fails!
        isa_ok($log_args[0], 'Mojo::Exception')
            or diag explain $log_args[0];
        # Do something with Mojo::Exception
    }

    sub is_error { 1 }
    sub is_debug { 0 }
}

{
    package Mojolicious::Test::Log; # an app that dies
    use Mojolicious::Lite;
    use Mojolicious::Plugin::Log::Any;

    plugin 'Log::Any';

    get '/error' => sub { die "Log Me" };
}

subtest test_exceptions => sub {
    my $t = Test::Mojo->new('Mojolicious::Test::Log');
    Log::Any::Adapter->set('Mojo::Test');

    # this succeeds, but the logging itself fails
    $t->get_ok('/error')
      ->status_is(500 => 'threw exception');
};

done_testing();
$ prove log.t
```

### Expected behavior

The test should pass.
That would mean that we're not stringifying a `Mojo::Exception`, and whatever log mechanism is in use can eg. extract a stack trace from the exception.

A real-life example of needing the exception _object_ is https://metacpan.org/release/Log-Any-Adapter-Sentry-Raven/source/lib/Log/Any/Adapter/Sentry/Raven.pm#L135

### Actual behavior
Instead, the exception is stringified by `Mojo::Log` (_before_ `emit` is called).
```
$ prove log.t
log.t ..
    #   Failed test 'The class (or class-like) '[error] [ab129e02] Log Me at log.t line 36.
    # ' isa 'Mojo::Exception''
    #   at /$path/Mojo/Log/Role/AttachLogger.pm line 71.
    #     The class (or class-like) '[error] [ab129e02] Log Me at log.t line 36.
    # ' isn't a 'Mojo::Exception'
    # [error] [b34e468f] Log Me at log.t line 36.
    # Looks like you failed 1 test of 3.
log.t .. 1/?
#   Failed test 'test_exceptions'
#   at log.t line 46.
# Looks like you failed 1 test of 1.
log.t .. Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests

Test Summary Report
-------------------
log.t (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
Files=1, Tests=1,  0 wallclock secs ( 0.01 usr  0.01 sys +  0.33 cusr  0.06 csys =  0.41 CPU)
Result: FAIL
```

### Suggested Fix
Prepend to the exception _message_ rather than the exception itself.
```
sub _log {
  my ($self, $level) = (shift, pop);
  my @msgs = ref $_[0] eq 'CODE' ? $_[0]() : @_;
+  if ($self->context) {
+    if (@msgs == 1 && blessed($msgs[0]) && $msgs[0]->isa('Mojo::Exception')) {
+      $msgs[0]->message("$self->{context} ".$msgs[0]->message);
+    }
+    else {
+      $msgs[0] = "$self->{context} $msgs[0]";
+    }
+  }
-  $msgs[0] = "$self->{context} $msgs[0]" if $self->{context};
  ($self->{parent} || $self)->emit('message', $level, @msgs);
}
```

**Comment by jrubinator** [comment]
Hi @kraih - would you be open to a PR for this?

**Comment by kraih** [comment]
This is not a bug. I'm not opposed to changes in behaviour, but they would have to be considered as a new feature. And especially not result in performance regressions.

**Comment by jrubinator** [comment]
Thanks @kraih. It looks to me like that behavior changed betweenhttps://metacpan.org/source/SRI/Mojolicious-8.23/lib/Mojo/Log.pm#L63 and https://metacpan.org/source/SRI/Mojolicious-8.24/lib/Mojo/Log.pm#L72. Was the stringification there for performance concerns?

```
8.24  2019-09-11
  - Added EXPERIMENTAL context method to Mojo::Log.
  - Added EXPERIMENTAL cleanup event to Mojo::IOLoop::Subprocess.
  - Added log helper to Mojolicious::Plugin::DefaultHelpers.
  - Improved log messages generated by Mojolicious to include request ids when
    possible.
```

**Comment by kraih** [comment]
It's been too long ago, i don't remember the details.

**Comment by jrubinator** [comment]
That makes sense.

I don't see any major performance changes from the suggested patch; it looks to me like the default behavior of `Mojo::Log` is going to be to stringify the message anyway:
* default: https://metacpan.org/source/SRI/Mojolicious-8.69/lib/Mojo/Log.pm#L63
* short: https://metacpan.org/source/SRI/Mojolicious-8.69/lib/Mojo/Log.pm#L87

So the only worry I can think of is that this is prohibitively nonperformant:
```
+    if (@msgs == 1 && blessed($msgs[0]) && $msgs[0]->isa('Mojo::Exception')) {
```

Are you worried about that too? Or should I proceed to submit this as a new feature?

**Comment by kraih** [comment]
Of course i am worried about that. The PR will need proof that there is no performance regression.

**Comment by jrubinator** [comment]
Are there any similar proofs from past PRs that I can use as a template? Benchmarking the additional if-statement on the linux machines available to me seems incomplete.

**Comment by jrubinator** [comment]
Well, here's what is does on my machine, for what it's worth:

*TL;DR: the timing fluctuates (I think related to the i/o costs of writing to a file), but I see no noticeable performance dip from the patch*

Script:
```
use strict;
use warnings;

{
    package Mojolicious::Benchmark;
    use Mojolicious::Lite;

    use Benchmark qw(timethese);
    use Mojo::Exception;
    use Mojo::File;

    get '/' => sub {
        my $c = shift;
        my $e = Mojo::Exception->new("Oh no!");
        my $o = bless {}, "Some::Object";
        timethese( 10_000, {
            "One message",  sub { $c->log->error("Message") },
            "Two messages", sub { $c->log->error("one", "two") },
            # reset, or this just keeps getting bigger
            "Exception",    sub { $c->log->error($e); $e->message("Oh no!") },
            "Object",       sub { $c->log->error($o) },
        });
        return $c->render(
            status => 200,
            text   => 'Benchmark complete',
        );
    }
}

use Test::Mojo;
use Test::More;

my $t = Test::Mojo->new('Mojolicious::Benchmark');
$t->get_ok('/');

done_testing;
```

Current behavior
```
$ MOJO_LOG_LEVEL=debug perl -Ilib timing.pl 2> out_orig.txt
Benchmark: timing 10000 iterations of Exception, Object, One message, Two messages...
 Exception: 12 wallclock secs ( 0.69 usr +  0.70 sys =  1.39 CPU) @ 7194.24/s (n=10000)
    Object: 11 wallclock secs ( 0.62 usr +  0.65 sys =  1.27 CPU) @ 7874.02/s (n=10000)
One message: 11 wallclock secs ( 0.63 usr +  0.71 sys =  1.34 CPU) @ 7462.69/s (n=10000)
Two messages: 12 wallclock secs ( 0.65 usr +  0.66 sys =  1.31 CPU) @ 7633.59/s (n=10000)
```

Behavior post proposed-patch:
```
$ MOJO_LOG_LEVEL=debug perl -Ilib timing.pl 2> out_new.txt
Benchmark: timing 10000 iterations of Exception, Object, One message, Two messages...
 Exception: 11 wallclock secs ( 0.67 usr +  0.65 sys =  1.32 CPU) @ 7575.76/s (n=10000)
    Object: 11 wallclock secs ( 0.59 usr +  0.62 sys =  1.21 CPU) @ 8264.46/s (n=10000)
One message: 11 wallclock secs ( 0.69 usr +  0.66 sys =  1.35 CPU) @ 7407.41/s (n=10000)
Two messages: 13 wallclock secs ( 0.67 usr +  0.86 sys =  1.53 CPU) @ 6535.95/s (n=10000)
```

@kraih is that sufficient due diligence, or is there something more I can verify first?

**Comment by jrubinator** [comment]
Realized that changing the underlying `Mojo::Exception` was probably a bad sign. Fixing that gives these numbers on rerun:

Before patch:
```
$ MOJO_LOG_LEVEL=debug perl -Ilib timing.pl 2> out_three.txt
Benchmark: timing 10000 iterations of Exception, Object, One message, Two messages...
 Exception: 12 wallclock secs ( 0.69 usr +  0.68 sys =  1.37 CPU) @ 7299.27/s (n=10000)
    Object: 12 wallclock secs ( 0.61 usr +  0.69 sys =  1.30 CPU) @ 7692.31/s (n=10000)
One message: 11 wallclock secs ( 0.58 usr +  0.69 sys =  1.27 CPU) @ 7874.02/s (n=10000)
Two messages: 13 wallclock secs ( 0.67 usr +  0.73 sys =  1.40 CPU) @ 7142.86/s (n=10000)
```
After patch:
```
$ MOJO_LOG_LEVEL=debug perl -Ilib timing.pl 2> out_non_damaging.txt
Benchmark: timing 10000 iterations of Exception, Object, One message, Two messages...
 Exception: 14 wallclock secs ( 0.85 usr +  0.69 sys =  1.54 CPU) @ 6493.51/s (n=10000)
    Object: 12 wallclock secs ( 0.67 usr +  0.72 sys =  1.39 CPU) @ 7194.24/s (n=10000)
One message: 12 wallclock secs ( 0.59 usr +  0.69 sys =  1.28 CPU) @ 7812.50/s (n=10000)
Two messages: 12 wallclock secs ( 0.61 usr +  0.70 sys =  1.31 CPU) @ 7633.59/s (n=10000)
```

So that would incur a small penalty on _exception_ logging. On the other hand, it seems unlikely to me that a tiny penalty on exception logging to acquire better logs is a big problem. In fact, I think we may actually take a larger one already, as evidenced by this one further benchmark:
```
"Generate",     sub { $c->log->error(Mojo::Exception->new("test")->trace->inspect) },
Generate: 13 wallclock secs ( 2.08 usr +  0.74 sys =  2.82 CPU) @ 3546.10/s (n=10000)
```

**Comment by kraih** [comment]
I believe this is resolved in 9.0.
